### ISSUE TITLE
DynamoDB: Global Table creation doesn't set retention policy for replica tables

### ISSUE THREAD

**Issue by BalintLab** [aws/aws-cdk#25740]
### Status

Investigating (Default)

### What is the issue?

The level 2 Table construct creates global tables that are not in alignment with DynamoDB best practices (uses `Table` CFN resource + custom resources vs the GlobalTable CFN resource)

This poses a serious risk of deleting the replicas if their retain policy was set up on the construct:
```
const XRTable2_lvl2 = new Table(this, 'Table', {
   tableName: "XRTable2_lvl2",
   partitionKey: { name: 'id', type: AttributeType.STRING },
   replicationRegions: ['us-east-1', 'us-east-2', 'us-west-2'],
   billingMode: BillingMode.PROVISIONED,
});
XRTable2_lvl2.autoScaleWriteCapacity({
   minCapacity: 1,
   maxCapacity: 10,
}).scaleOnUtilization({ targetUtilizationPercent: 75 });
XRTable2_lvl2.applyRemovalPolicy(RemovalPolicy.RETAIN)
```

I used the above code to create my global tables. When I deleted the CF stack manually, root table was preserved as per removalpolicy, but all replicas were deleted.

This construct should adhere to dynamodb best practices and create a GlobalTable resource: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/V2globaltables_reqs_bestpractices.html

**Reference:**
- Documentation: https://github.com/aws/aws-cdk/tree/main/packages/aws-cdk-lib/aws-dynamodb#amazon-dynamodb-global-tables

- Code: https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-dynamodb/lib/replica-provider.ts

### What is the impact?

If Global tables are created this way, the replica tables retention policy won't be set. This means, the replica tables can be accidentally deleted despite what the code snippet suggests

### Who is affected?

Anyone creating global tables based on this doc using this library

### How do I resolve this?

Customers could use the level 1 CFNGlobalTable constuct as a workaround: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_dynamodb.CfnGlobalTable.html

**Comment by peterwoodworth** [comment]
The replicas aren't created through Cfn, so this would need to be handled by the custom resource handler. Could potentially skip deleting some regions here https://github.com/aws/aws-cdk/blob/bbdb16ad65b1217901f92782a74a6f13222cd684/packages/aws-cdk-lib/aws-dynamodb/lib/replica-handler/index.ts#L31-L43

**Comment by BalintLab** [comment]
I'd also recommend creating a global table resource (`CfnGlobalTable`) when `replicaregions` is set as per DynamoDB best practices

**Comment by rix0rrr** [comment]
This issue was for the existing `Table` construct, which used custom resources to implement table replication. We no longer recommend the use of the `Table` construct.

Instead, the `TableV2` construct has been released in 2.95.1 (#27023) which maps to the `AWS::DynamoDB::GlobalTable` resource, has better support for replication and does not suffer from the issue described here. 

Be aware that there are additional deployment steps involved in a migration from `Table` to `TableV2`. You need to do a RETAIN deployment, a delete deployment, then change the code to use `TableV2` and then use `cdk import`. A link to a full guide will be posted once it is available. 

Here are some other resources to get you started (using `CfnGlobalTable` instead of `TableV2`) if you want to get going on the migration:

* [Migrating from Table to CfnGlobalTable](https://github.com/aws-samples/amazon-dynamodb-table-to-global-table-cdk), by Samaneh Utter, an AWS Senior Solutions Architect
* [AWS CDK and Amazon DynamoDB Global Tables](https://dev.to/aws-builders/aws-cdk-and-amazon-dynamodb-global-tables-3p0b), by Wojciech Matuszewski from Stedi
* [Migrating from Table to CfnGlobalTable](https://medium.com/@sharon.ad94/effortlessly-migrate-dynamodb-table-to-global-table-with-cloudformation-or-cdk-no-downtime-no-a1f62093d420), by Sharon Diskin from CloudBit
